aptos-move/aptos-vm/src/sharded_block_executor/sharded_executor_client.rs

```diff
@@ -0,0 +1,197 @@
+            message_txs: Arc::new(message_txs.into_iter().map(Mutex::new).collect()),
+        }
+    }
+
+    pub fn create_sharded_executor_clients(
+        num_shards: usize,
+        num_threads: Option<usize>,
+    ) -> Vec<Self> {
+        let num_threads = num_threads
+            .unwrap_or_else(|| (num_cpus::get() as f64 / num_shards as f64).ceil() as usize);
+        let mut cross_shard_msg_txs = vec![];
+        let mut cross_shard_msg_rxs = vec![];
+        for _ in 0..num_shards {
+            let (messages_tx, messages_rx) = std::sync::mpsc::channel();
+            cross_shard_msg_txs.push(messages_tx);
+            cross_shard_msg_rxs.push(messages_rx);
+        }
+        cross_shard_msg_rxs
+            .into_iter()
+            .enumerate()
+            .map(|(shard_id, rx)| {
+                Self::new(
+                    shard_id as ShardId,
+                    num_threads,
+                    cross_shard_msg_txs.clone(),
+                    rx,
+                )
+            })
+            .collect()
+    }
+
+    fn create_cross_shard_state_view<'a, S: StateView + Sync + Send>(
+        &self,
+        base_view: &'a S,
+        sub_block: &SubBlock<Transaction>,
+    ) -> CrossShardStateView<'a, S> {
+        let mut cross_shard_state_key = HashSet::new();
+        for txn in &sub_block.transactions {
+            for (_, storage_locations) in txn.cross_shard_dependencies.required_edges_iter() {
+                for storage_location in storage_locations {
+                    cross_shard_state_key.insert(storage_location.clone().into_state_key());
+                }
+            }
+        }
+        CrossShardStateView::new(self.shard_id, cross_shard_state_key, base_view)
+    }
+
+    fn execute_sub_block<S: StateView + Sync + Send>(
+        &self,
+        sub_block: SubBlock<Transaction>,
+        round: usize,
+        state_view: &S,
+        concurrency_level: usize,
+        maybe_block_gas_limit: Option<u64>,
+    ) -> Result<Vec<TransactionOutput>, VMStatus> {
+        trace!(
+            "executing sub block for shard {} and round {}",
+            self.shard_id,
+            round
+        );
+        let cross_shard_commit_sender = CrossShardCommitSender::new(
+            self.shard_id,
+            self.message_txs
+                .iter()
+                .map(|t| t.lock().unwrap().clone())
+                .collect(),
+            &sub_block,
+        );
+
+        let (callback, callback_receiver) = oneshot::channel();
+
+        let message_rxs = self.message_rx.clone();
+        let self_message_tx = Arc::new(Mutex::new(
+            self.message_txs[self.shard_id].lock().unwrap().clone(),
+        ));
+        let cross_shard_state_view =
+            Arc::new(self.create_cross_shard_state_view(state_view, &sub_block));
+        let cross_shard_state_view_clone1 = cross_shard_state_view.clone();
+        self.executor_thread_pool.scope(|s| {
+            s.spawn(move |_| {
```

Comment:
Here 1 thread is occupied. If the caller gives `concurrency_level==2`, then the block-stm will get only 1 thread and then stuck?...

Reply:
Thanks for confirming, finally found the "+ 2"...